Topics/Tech_Stacks/PostgreSQL_psycopg2.md

```diff
@@ -45,85 +46,252 @@ A quick and basic runover of PostgreSQL.
 ```
-    SELECT col1_table1, col2_table2 FROM table1_name, table2_name;
+SELECT * FROM sports, teams;
 ```
-This operation, first, performs a cross product between table1 and table2, where every row of one table is matched with every other row of the other table. Then it selects the data present in the new cross-products table from columns col1_table1 and col2_table2.
+In this query, each row of table `sports` is combined with each row of the table `teams`, resulting in a combination of all rows:
+| sport_id | sport_name | team_id | team_name | sport_id |
+| -------- | ---------- | ------- | --------- | -------- |
+| 1 | basketball | 1 | Toronto Raptors | 1 |
+| 1 | basketball | 2 | San Antonio Spurs | 1 |
+| 1 | basketball | 3 | Golden State Warriors | 1 |
+| 1 | basketball | 4 | Manchester City F.C. | 2 |
+| 1 | basketball | 5 | Paris Saint-Germain F.C. | 2 |
+| 2 | soccer | 1 | Toronto Raptors | 1 |
+| 2 | soccer | 2 | San Antonio Spurs | 1 |
+| 2 | soccer | 3 | Golden State Warriors | 1 |
+| 2 | soccer | 4 | Manchester City F.C. | 2 |
+| 2 | soccer | 5 | Paris Saint-Germain F.C. | 2 |
+| 3 | baseball | 1 | Toronto Raptors | 1 |
+| 3 | baseball | 2 | San Antonio Spurs | 1 |
+| 3 | baseball | 3 | Golden State Warriors | 1 |
+| 3 | baseball | 4 | Manchester City F.C. | 2 |
+| 3 | baseball | 5 | Paris Saint-Germain F.C. | 2 |
+
+However, not all rows of the output are meaningful. For example, the fourth row associates basketball with Manchester City F.C., which is a soccer team. To filter out the nonsensical combinations, we can use the `WHERE` clause:
+```
+SELECT * FROM sports, teams WHERE sports.sport_id = teams.sport_id;
+```
+This clause allows you to retrieve only the rows that meet the specified conditions. The output of this query would be:
+| sport_id | sport_name | team_id | team_name | sport_id |
+| -------- | ---------- | ------- | --------- | -------- |
+| 1 | basketball | 1 | Toronto Raptors | 1 |
+| 1 | basketball | 2 | San Antonio Spurs | 1 |
+| 1 | basketball | 3 | Golden State Warriors | 1 |
+| 2 | soccer | 4 | Manchester City F.C. | 2 |
+| 2 | soccer | 5 | Paris Saint-Germain F.C. | 2 |
+
+Although the query is now more targeted and relevant, the two `sport_id` columns are redundant. To get rid of the redundancy, one solution is to change the columns in the `SELECT` clause so that only one of the `sport_id` columns is retrieved:
+```
+SELECT sports.sport_id, sport_name, team_id, team_name FROM sports, teams WHERE sports.sport_id = teams.sport_id;
+```
+Another option is to use `NATURAL JOIN` in the `FROM` clause:
+```
+SELECT * FROM sports NATURAL JOIN teams;
+```
+`NATURAL JOIN` automatically matches and combines the columns with the same name in the tables being joined. Since `sport_id` is an attribute in both `sports` and `teams`, the resulting set includes all unqiue columns from both tables and excludes duplicate columns. The output of these queries are:
```

Comment:
Typo: "... includes all unique* columns from both ..."

Reply:
Thanks for pointing it out